### PR TITLE
Go taxon constraint

### PR DESCRIPTION
**Pull request by laisecf**

### CODE REVIEW

File: pyinterprod/pronto/goa.py [hunk 1]
```diff
@@ -234,3 +235,82 @@ def _get_constraints(term_id: str, ancestors: dict, constraints: dict) -> set:
...
+
+        sql = """
+                INSERT INTO GO2CONSTRAINTS (go_id, relationship, taxon)
+                VALUES (%s, %s, %s)
+                """
+
+        records = []
+        for go_id, relat2const in go2constraints.items():
+            for relation, constraints in relat2const.items():
+                for constr_id in constraints:
```
Comment (laisecf) [inline pyinterprod/pronto/goa.py:305]: About these "for", it's bothering me to have an n^3 despite having very few elements in the internal "for"s. I'm thinking about how to refactor the creation of this dict to improve here.

File: pyinterprod/pronto/goa.py [hunk 2]
```diff
@@ -234,3 +234,82 @@ def _get_constraints(term_id: str, ancestors: dict, constraints: dict) -> set:
...
+    ora_cur.execute(
+        """
+        SELECT UNION_ID, LISTAGG(TAXON_ID, ',')
+        FROM GO.NCBI_TAXON_UNIONS
+        GROUP BY UNION_ID
+        """
+    )
+    union2taxon = {}
+    for union_id, taxon_ids in ora_cur:
+        union2taxon[union_id] = taxon_ids.split(",")
```
Comment (matthiasblum) [inline pyinterprod/pronto/goa.py:253]: Cast directly to a `set`:

```suggestion
        union2taxon[union_id] = set(taxon_ids.split(","))
```

File: pyinterprod/pronto/goa.py [hunk 3]
```diff
@@ -234,3 +234,82 @@ def _get_constraints(term_id: str, ancestors: dict, constraints: dict) -> set:
...
+
+        if relationship not in go2constraints[go_id]:
+            go2constraints[go_id][relationship] = set()
+
+        if tax_id_type == "NCBITaxon_Union":
+            current_tax_id = set(union2taxon[tax_id])
+        elif tax_id_type == "NCBITaxon":
+            current_tax_id = {tax_id}
+        else:
+            raise f"Unknown TAX_ID_TYPE: {tax_id_type}"
```
Comment (matthiasblum) [inline pyinterprod/pronto/goa.py:279]: Exceptions must derive from one of the [base exceptions](https://docs.python.org/3/library/exceptions.html) (in this case `ValueError` seems to be the best fit):

```suggestion
            raise ValueError(f"Unknown TAX_ID_TYPE: {tax_id_type}")
```

File: pyinterprod/pronto/goa.py [hunk 4]
```diff
@@ -234,3 +234,82 @@ def _get_constraints(term_id: str, ancestors: dict, constraints: dict) -> set:
...
+        GROUP BY UNION_ID
+        """
+    )
+    union2taxon = {}
+    for union_id, taxon_ids in ora_cur:
+        union2taxon[union_id] = taxon_ids.split(",")
+
+    ora_cur.execute(
+        """
+        SELECT T.GO_ID, AT.GO_ID, TC.RELATIONSHIP, TC.TAX_ID_TYPE, TC.TAX_ID
```
Comment (matthiasblum) [inline pyinterprod/pronto/goa.py:257]: No need to select columns we don't use:

```suggestion
        SELECT T.GO_ID, TC.RELATIONSHIP, TC.TAX_ID_TYPE, TC.TAX_ID
```

File: pyinterprod/pronto/goa.py [hunk 5]
```diff
@@ -234,3 +234,82 @@ def _get_constraints(term_id: str, ancestors: dict, constraints: dict) -> set:
...
+        if go_id not in go2constraints:
+            go2constraints[go_id] = {}
+
+        if relationship not in go2constraints[go_id]:
+            go2constraints[go_id][relationship] = set()
+
+        if tax_id_type == "NCBITaxon_Union":
+            current_tax_id = set(union2taxon[tax_id])
+        elif tax_id_type == "NCBITaxon":
+            current_tax_id = {tax_id}
```
Comment (matthiasblum) [inline pyinterprod/pronto/goa.py:277]: Why not adding directly to `go2constraints[go_id][relationship]`?

File: pyinterprod/pronto/goa.py [hunk 6]
```diff
@@ -234,3 +234,82 @@ def _get_constraints(term_id: str, ancestors: dict, constraints: dict) -> set:
...
+                relationship VARCHAR(50) NOT NULL,
+                taxon INTEGER NOT NULL
+            )
+            """
+        )
+
+        sql = """
+                INSERT INTO GO2CONSTRAINTS (go_id, relationship, taxon)
+                VALUES (%s, %s, %s)
+                """
```
Comment (matthiasblum) [inline pyinterprod/pronto/goa.py:299]: Indentation:

```suggestion
        sql = """
            INSERT INTO GO2CONSTRAINTS (go_id, relationship, taxon)
            VALUES (%s, %s, %s)
        """
```

File: pyinterprod/pronto/goa.py [hunk 7]
```diff
@@ -234,3 +234,82 @@ def _get_constraints(term_id: str, ancestors: dict, constraints: dict) -> set:
...
+
+        records = []
+        for go_id, relat2const in go2constraints.items():
+            for relation, constraints in relat2const.items():
+                for constr_id in constraints:
+                    records.append((go_id, relation, constr_id))
+            if len(records) == 1000:
+                pg_cur.executemany(sql, records)
+                pg_con.commit()
+                records.clear()
```
Comment (matthiasblum) [inline pyinterprod/pronto/goa.py:309]: You need to check for the number of cached records after each `append()`:

```suggestion
                    records.append((go_id, relation, constr_id))

                    if len(records) == 1000:
                        pg_cur.executemany(sql, records)
                        pg_con.commit()
                        records.clear()
```

**Review by matthiasblum — CHANGES_REQUESTED** [review]
Good overall, with a few issues to resolve.

Also, a few variables could have better names, e.g.

- `union2taxon` should be renamed for consistency with the data it contains, such as `union2taxa` (plural of _taxon_ is _taxa_, and a union is not for _one taxon_)
- `constr_id` is a taxon ID, so better name it `tax_id`, or `taxon_id`.

File: pyinterprod/pronto/goa.py [hunk 8]
```diff
@@ -234,3 +234,78 @@ def _get_constraints(term_id: str, ancestors: dict, constraints: dict) -> set:
...
+        if relationship not in go2constraints[go_id]:
+            go2constraints[go_id][relationship] = set()
+
+        if tax_id_type == "NCBITaxon_Union":
+            go2constraints[go_id][relationship] |= set(union2taxa[tax_id])
+        elif tax_id_type == "NCBITaxon":
+            go2constraints[go_id][relationship] = {tax_id}
+        else:
+            raise ValueError(f"Unknown TAX_ID_TYPE: {tax_id_type}")
+
```
Comment (matthiasblum) [inline pyinterprod/pronto/goa.py:280]: Need to close the Oracle cursor/connection.

File: pyinterprod/pronto/goa.py [hunk 9]
```diff
@@ -234,3 +234,78 @@ def _get_constraints(term_id: str, ancestors: dict, constraints: dict) -> set:
         result |= _get_constraints(parent_id, ancestors, constraints)
 
     return result
+
+
+def import_go_constraints(go_url: str, pg_url: str):
+    pg_con = psycopg.connect(**url2dict(pg_url))
```
Comment (matthiasblum) [inline pyinterprod/pronto/goa.py:240]: The PostgreSQL connection should be opened later, when we need it (just before `with pg_con.cursor() as pg_cur:`).

File: pyinterprod/pronto/goa.py [hunk 10]
```diff
@@ -234,3 +234,78 @@ def _get_constraints(term_id: str, ancestors: dict, constraints: dict) -> set:
...
+                for taxon_id in constraints:
+                    records.append((go_id, relation, taxon_id))
+                    if len(records) == 1000:
+                        pg_cur.executemany(sql, records)
+                        pg_con.commit()
+                        records.clear()
+        if records:
+            pg_cur.executemany(sql, records)
+            pg_con.commit()
+            records.clear()
```
Comment (matthiasblum) [inline pyinterprod/pronto/goa.py:311]: Need to:
- Create an index on `go_id`
- Close the PostgreSQL connection

**Review by matthiasblum — CHANGES_REQUESTED** [review]
`import_go_constraints()` is not called from anywhere. You need to add a task in [`get_pronto_tasks()`](https://github.com/ProteinsWebTeam/pyinterprod/blob/ae64c0458ef225dd1b2888c7656318456ccd2d15/pyinterprod/cli.py#L13) so constraints are imported when we refresh the Pronto database.

You could rename the `annotations` task to `go-terms`, and name the new task `go-constraints`. The new task does not depend on any other task.

**Review by matthiasblum — APPROVED** [review]
LGTM :rocket:

**Pull request merged**